### PR TITLE
[fix] output of `embedding_bag` with non-contiguous weight 

### LINKED ISSUE #43723: embedding_bag running on CPU produces wrong result when weight tensor is non-contiguous.
## 🐛 Bug

There's a bug in PyTorch's torch.nn.functional.embedding_bag that is still in the PyTorch 1.6.0 (an earlier version I checked is 1.3.1, where the problem also exists). Specifically, if the weights are not contiguous, the results will be wrong (without a warning). AFAIC this only happens on CPU. Does not appear to happen on CUDA.

## To Reproduce

```python
import torch
import torch.nn.functional as F
print(torch.__version__)

weight_tensor = torch.tensor(
    [[0., 1., 2., 3.],
    [4., 5., 6., 7.],
    [8., 9., 10., 11.]]
)

weight_tensor_non_contig = weight_tensor[:, :3]  # This is non-contiguous strided.
print(f"weight_tensor_non_contig stride: {weight_tensor_non_contig.stride()}")

weight_tensor_contig = weight_tensor_non_contig.clone().contiguous()  # Contig-strided.
print(f"weight_tensor_contig stride: {weight_tensor_contig.stride()}")

index = torch.tensor([0, 1, 2])
offsets = torch.tensor([0, 2])
# Run embedding bag. What this should do is take segment mean, where the segments are
# defined by the offsets. In this case, 0:2 is one segment, 2:3 is one segment
output_non_contig = F.embedding_bag(
    input=index,
    weight=weight_tensor_non_contig,
    offsets=offsets,
    mode="mean",
)
output_contig = F.embedding_bag(
    input=index,
    weight=weight_tensor_contig,
    offsets=offsets,
    mode="mean",
)
print(f"Input Tensor:\n{iweight_tensor_contig}")

# The outputs returned in the non-contiguous version is wrong.
print(f"Non-contigous result (wrong):\n{output_non_contig}")

# The outputs returned in the contiguous version is correct.
print(f"Contigous result   (correct):\n{output_contig}")
```
Returns:
```text
1.6.0
input_tensor_non_contig stride: (4, 1)
input_tensor_contig stride: (3, 1)
Input Tensor:
tensor([[ 0.,  1.,  2.],
        [ 4.,  5.,  6.],
        [ 8.,  9., 10.]])
Non-contigous result (wrong):
tensor([[1.5000, 2.5000, 3.5000],
        [6.0000, 7.0000, 8.0000]])
Contigous result   (correct):
tensor([[ 2.,  3.,  4.],
        [ 8.,  9., 10.]])
```
## Environment

PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Quadro RTX 5000
GPU 1: Quadro RTX 5000

Nvidia driver version: 410.78
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] numpydoc==0.9.2
[pip3] torch==1.6.0
[pip3] torchdiffeq==0.0.1
[pip3] torchvision==0.6.0a0+35d732a
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2020.0                      166
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.15           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] numpy                     1.18.1           py37h4f9e942_0
[conda] numpy-base                1.18.1           py37hde5b4d6_1
[conda] numpydoc                  0.9.2                      py_0
[conda] torch                     1.6.0                    pypi_0    pypi
[conda] torchdiffeq               0.0.1                    pypi_0    pypi
[conda] torchvision               0.5.0+cu100              pypi_0    pypi

## Additional context
This problem seems to not exist on the CUDA version.

cc @ezyang @gchanan @zou3519 @VitalyFedyunin @albanD @mruberry

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #43723 

use weight.contiguous on fast-path as it expects contiguous tensor.

TODO:
* [x] Add tests

### CODE REVIEW

File: aten/src/ATen/native/EmbeddingBag.cpp
Comment (glaringlee) [inline aten/src/ATen/native/EmbeddingBag.cpp:94]: nit: the two lines can be merged into one.
src.contiguous().data_ptr<float>()

File: aten/src/ATen/native/EmbeddingBag.cpp
Comment (glaringlee) [inline aten/src/ATen/native/EmbeddingBag.cpp:223]: same here

**Review by glaringlee — COMMENTED** [review]
Thanks for adding tests.
I have a nit comment, please take a look at.

File: aten/src/ATen/native/EmbeddingBag.cpp
Comment (kshitij12345) [reply aten/src/ATen/native/EmbeddingBag.cpp:94]: Makes sense! Thanks!

**Comment by glaringlee** [comment]
@kshitij12345 Thanks a lot for fixing this. i will approve this once the CI test is done without issues.

**Review by glaringlee — APPROVED** [review]
LGTM now.

**Comment by kshitij12345** [comment]
@glaringlee Gentle Ping :)

**Comment by glaringlee** [comment]
@kshitij12345 
Hi, your code will be landed today. 
The code is in landing process, here is what happened, after the approval on github, I imported the code to our fb internal system, and there will be another round approval there before landing the code. We were on time off since last Friday due to labor day, so the other approval will happen today and your code will be landed then. Thanks for reaching out.

**Comment by kshitij12345** [comment]
@glaringlee Oops I forgot it was Federal Holiday. Thanks!

**Comment by glaringlee** [comment]
@kshitij12345 
Hi, can you please do a rebase to your code?

File: test/test_nn.py
Comment (ngimel) [inline test/test_nn.py:11165]: this is contiguous tensor

Comment (kshitij12345) [reply test/test_nn.py:11165]: Right! Great catch!!
Was supposed to be 
```python
weight_tensor = torch.randn(3, 4, dtype=dtype, device=device)
```

Comment (glaringlee) [reply test/test_nn.py:11165]: @kshitij12345 I'll put a fix for u.

Comment (kshitij12345) [reply test/test_nn.py:11165]: Sure. Btw I am free this evening so I can put it up as well. Let me know if I should.

Comment (glaringlee) [reply test/test_nn.py:11165]: Ah, no worries. I have a clean pytorch repo on hand, will patch this shortly.

**Comment by glaringlee** [comment]
#44382 is entered for patching this.

File: aten/src/ATen/native/EmbeddingBag.cpp
Comment (dzhulgakov) [inline aten/src/ATen/native/EmbeddingBag.cpp:93]: wait, this is still a bug, right? If the tensor is non-contiguous then `contiguous()` will return a new tensor and it will be immediately destroyed (because we don't keep a reference to it around). So `src_data` will point to the deallocated memory :(

I wonder why ASAN doesn't catch it.

It should be
```
auto src_contig = src.contiguous();
auto* src_data = src_contig.data_ptr<float>();
```

cc @glaringlee @ngimel

Comment (glaringlee) [reply aten/src/ATen/native/EmbeddingBag.cpp:93]: @dzhulgakov oh, shoot.........my bad, will fix soon.

**Pull request closed without merging**
